[Blog] Github action CI/CD (#110)
JankariTech/blog: src/assets/githubAction/githubActionBeginner.md

---
title: Github Actions - CI & CD
authorName: Nabin Ale
authorAvatar : https://avatars.githubusercontent.com/u/61624650?s=400&u=8f992a9ca189af19c40cb070a5af1e02669d3884&v=4
authorLink: https://github.com/nabim777
createdAt: Oct 30, 2023
tags: CI, CD, githubAction
banner: https://raw.githubusercontent.com/JankariTech/blog/master/src/imgs/fallback_banner.png
---
GitHub Actions is an automation and CI/CD service provided by GitHub. It allows us to automate our software development process or workflows, from building and testing code to deploying it to different environments. It's a really effective tool that can both increase software quality and save us time.

## Features in github actions:
- It allows you to automate your Software Development Life Cycle (SDLC) processes or workflows.
- It simplifies CI/CD integration, enabling seamless DevOps practices.

> ## Components of github actions:

Reviewer: ```suggestion
## Components of GitHub Actions:
```

1. Workflows
2. Events
3. Jobs
4. Steps
5. Actions
6. Runners

![components](/src/assets/githubAction/images/components_of_github_action.png)

### 1. Workflows:
Workflows are automated procedures that you add to your repository. Workflows will run when triggered by an event in your repository, or they can be triggered manually, or at a defined schedule. It can be used to build, test, package, release, or deploy a project on Github.

### 2. Events:
Events are specific activities like code pushes, pull requests creation, issue creation, issue comments, etc. that triggers a workflow.

### 3. Jobs

> Jobs are the set of steps that were executed in same runner. A workflow with multiple jobs will run in parallel which is default, but it can be configured so that jobs will run sequentially.

Reviewer: ```suggestion
Jobs are the set of steps that are executed in the same runner. A workflow with multiple jobs will run in parallel which is the default, but it can be configured so that jobs will run sequentially.
```

### 4. Steps
A step is an individual task that can run the commands in a job. These can be actions or commands.

### 5. Actions

> Actions are the standalone commands for GitHub Actions platform that perform a complex but frequently repeated task. Actions helps to reduce the repetitive code in workflows.

Reviewer: ```suggestion
Actions are the standalone commands for the GitHub Actions platform that perform a complex but frequently repeated task. Actions help to reduce the repetitive code in workflows.
```

### 6. Runner

> Runner is a github actions server where our workflows run. A runner listens for available jobs. Each runner runs jobs at a time, and reports the progress, logs, and results back to github. Github hosted runners are based on Ubuntu, Linux, Microsoft Windows, and macos.

Reviewer: ```suggestion
A runner is a GitHub actions server where our workflows run. A runner listens for available jobs. Each runner runs jobs at a time, and reports the progress, logs, and results back to GitHub. GitHub hosted runners are based on Ubuntu, Linux, Microsoft Windows, and Mac OS.
```

"Each runner runs jobs at a time" - what do you want to say - "1 job at a time", "multiple jobs at a time"?

Author: I meant to say "multiple jobs at a time" :sweat_smile:

Note:
One event can trigger many workflows, a workflow can contain many jobs, and a job can contain many steps.

### Example:

> ![process_in_githubActions](/src/assets/githubAction/images/process.png)

Reviewer 2: It's usually good to have a short caption that describes the image

> ### Getting Started:

Reviewer 2: ```suggestion
## Getting started:
```
--> check the title hierarchy, it seems there are too many H3 titles. I think this item is no longer part of the previous H2 (Components of GitHub Actions:)

> Before we dive into an example, make sure we have a repository where we want to set up your CI/CD workflow. If we don't have one, create a new repository for example named `action-hero`.

Reviewer 3: ```suggestion
Before we dive into an example, make sure you have a repository where you want to set up your CI/CD workflow. If you don't have one, create a new repository for example named `action-hero`.
```

Step 1: Clone your repository into your system
```bash
git clone <your_github_repo_url>
```

Step 2: Go to path `action-hero` and make a folder `.github`
```bash
cd action-hero
mkdir .github
```

Step 4: Then, go to path `.github` and make a folder `workflows`
```bash
cd .github
mkdir workflows
```

Step 5: Inside the folder `workflows` create a `hello_world.yml` file
```bash
cd workflows
touch hello_world.yml
```
Our file structure will look like this
```
📦Action-hero
┗ 📂.github
┃ ┣ 📂workflows
┃ ┃ ┗ 📜hello_world.yml
```

Add the following code snippet to your `hello_world.yml` file:

```
name: Hello world workflow

on:
  push:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    - name: Print a message
      run: echo "Hello, GitHub Actions!"
```
In this example:
- The workflow is triggered by a push event on the main branch.
- The job named "build" runs on the latest version of Ubuntu.
- The first step checks out the repository using the actions/checkout action.
- The second step simply prints a message using the echo command.

## GitHub Actions activity:

> ![alt text](/src/assets/githubAction/images/output.png)

Reviewer 2: it's a good practice to define an alt text to increase accessibility

## Conclusion:

> In this vlog, we've explored GitHub Actions, an automation and CI/CD service provided by GitHub. We've learned about its key components and how they work together to automate software development workflows.

Reviewer 4: isn't a vlog a video-blog?

Author: My mistake :sweat_smile:

Reviewer 2: I like the idea of doing some video blogs :)

GitHub Actions enables us to automate tasks like building, testing, and deploying code, leading to improved software quality and time savings.

With a practical example, we've seen how easy it is to set up a basic workflow.

> In summary, GitHub Actions is a powerful tool for speedy development processes, making it easier to deliver high-quality software efficiently.

Reviewer 2: that's great. If you're planning to write more on this topic you could already add a teaser for your next blog.

Author: Currently, I haven't made quite plans for another blog, so  I can't promise a teaser for the next one just yet. But I'll definitely keep your suggestion in mind if I decide to write more on this topic.
 
Would it be fine if I add later?
Or 
Is it not good practice so I need to add it now? @koebel @phil-davis

Reviewer 2: sure, you can update your blog post once you're working on a more advance one on this topic or once you have published another one. 
I agree that it's better not to promise something to the reader that might not happen in the near future.

---
title: GitHub Actions - CI & CD
authorName: Nabin Ale
authorAvatar : https://avatars.githubusercontent.com/u/61624650?s=400&u=8f992a9ca189af19c40cb070a5af1e02669d3884&v=4
authorLink: https://github.com/nabim777
createdAt: Oct 30, 2023
tags: CI, CD, github, githubAction
banner: https://blog.jankaritech.com/src/assets/githubAction/images/githubaction_banner.png
---
GitHub Actions is an automation and CI/CD service provided by GitHub, allowing us to automate our software development processes or workflows, from building and testing code to deploying it to different environments. It's an effective tool that can enhance software quality and save time.

## Some features of GitHub Actions:
- It allows you to automate your Software Development Life Cycle (SDLC) processes or workflows.
- It simplifies CI/CD integration, facilitating seamless DevOps practices.

## Components of GitHub Actions:
1. Workflows
2. Events
3. Jobs
4. Steps
5. Actions
6. Runners

![components](/src/assets/githubAction/images/components_of_github_action.png)

### 1. Workflows:

> Workflows are automated procedures that you add to your repository. It will run when triggered by an event in your repository, or they can be triggered manually, or at a defined schedule. It can be used to build, test, package, release, or deploy a project on GitHub.

Reviewer 5: The second version ("They will run...") is more consistent with the plural subject "workflows" and is preferable in this context. It maintains grammatical agreement and clarity throughout the passage and I agree with Koebel on this

### 5. Actions
Actions are the standalone commands for the GitHub Actions platform that perform complex but frequently repeated tasks. Actions help to reduce repetitive code in workflows.

### 6. Runner
A runner is a GitHub Actions server where our workflows run. It listens for available jobs, runs multiple jobs at a time, and reports the progress, logs, and results back to GitHub. GitHub hosted runners are based on Ubuntu, Linux, Microsoft Windows, and Mac OS.

Note:
One event can trigger many workflows, a workflow can contain many jobs, and a job can contain many steps.

### Process in GitHub Actions:

![process_in_githubActions](/src/assets/githubAction/images/process.png)

## Getting started:

Before we dive into an example, make sure you have a repository where you want to set up your CI/CD workflow. If you don't have one, create a new repository, for example, named `action-hero`.

Step 1: Clone your repository into your system
```bash
git clone <your_github_repo_url>
```

Step 2: Go to path `action-hero` and create a folder named `.github`
```bash
cd action-hero
mkdir .github
```

> Step 3: Then, go to path `.github` and create a folder named `workflows`

Reviewer 2: ```suggestion
Step 3: Go to path `.github` and create a folder named `workflows`
```
"Then" is not needed since it's the only thing your telling the reader to do in this step

```bash
cd .github
mkdir workflows
```

> Step 4: Inside the folder `.github`, create a subfolder named `workflows`

Reviewer: ```suggestion
Step 4: Inside the folder `workflows`, create a file named `hello_world.yml`
```

```bash
cd workflows
touch hello_world.yml
```

> Our file structure should look like this

Reviewer 2: ```suggestion
Our file structure should look like this:
```

```
📦Action-hero
┗ 📂.github
┃ ┣ 📂workflows
┃ ┃ ┗ 📜hello_world.yml
```

> Step 5: Add the following code snippet to your file `hello_world.yml`:

Reviewer 2: ```suggestion
Step 5: Add the following code snippet to your `hello_world.yml` file:
```

```
name: Hello world workflow

on:
  push:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    - name: Print a message
      run: echo "Hello, GitHub Actions!"
```

In this example:
- The workflow is triggered by a push event on the main branch.
- The job named "build" runs on the latest version of Ubuntu.
- The first step checks out the repository using the `actions/checkout` action.
- The second step simply prints a message using the `echo` command.

## GitHub Actions activity:

![github activity](/src/assets/githubAction/images/output.png)

> The image above is an output from a GitHub Actions workflow run. It shows that the workflow was successful and completed the `Hello world workflow` job, which consisted of two steps: `Checkout repository` and `Print a message`.

Reviewer 2: ```suggestion
The image above is the output from our GitHub Actions workflow run. It shows that the workflow was successful and completed the `Hello world workflow` job, which consisted of two steps: `Checkout repository` and `Print a message`.
```

Since you're referring to a specific workflow, not just any one, it's better to use a specific article (the) instead of an unspecific article (a).

The `Print a message` step executed the command `echo "Hello, GitHub Actions!"`, resulting in the output `Hello, GitHub Actions!` being displayed in the workflow log.

> ## Conclusion:

Reviewer 2: ```suggestion
## Conclusion
```

---

> title: GitHub Actions - CI & CD

Author: ```suggestion
title: GitHub Actions - CI & CD (For Beginners)
```
I think I have to change the topics since its for beginner
Or 
It will be ok 
Or 
need suggestion

Reviewer 2: That's fine. Maybe ask your colleagues for suggestions too :)

You could also say something like "Introduction to GitHub Actions - CI & CD"

And here are some ideas that an AI based title generator tool gave me :)
1. "Mastering GitHub Actions: A Beginner's Guide to CI & CD"
2. "Unleashing the Power of GitHub Actions: A Step-by-Step CI & CD Tutorial for Beginners"
3. "Demystifying GitHub Actions: The Ultimate CI & CD Tool for Beginners"
4. "Boost Your Development Workflow with GitHub Actions: A Beginner's Introduction to CI & CD"
5. "GitHub Actions Made Easy: A Beginner's Guide to Implementing CI & CD in Your Projects"

Reviewer: AI made much longer titles than I thought of ;)

Reviewer 2: I think AI can be useful to get more ideas, but some of these AI generated titles might not match with the content of your blog post. Also decide on a title that you feel comfortable with and that uses words you would also be using.

authorName: Nabin Ale

> authorAvatar : https://avatars.githubusercontent.com/u/61624650?s=400&u=8f992a9ca189af19c40cb070a5af1e02669d3884&v=4

Reviewer 6: ```suggestion
authorAvatar : https://avatars.githubusercontent.com/u/61624650?v=4
```
Short link is enough

authorLink: https://github.com/nabim777
createdAt: Oct 30, 2023

> tags: CI, CD, github, githubAction

Reviewer 7: ```suggestion
tags: CI, CD, github, Github Actions
```

Author: I think spaces are not included in tags

Author: @SagarGi   I think we don't use space in between for tag according to the other site tag I have seen example:fb, youtube, twitter, instagram?

> tags: CI, CD, github, githubAction

Reviewer 6: ```suggestion
tags: continuous integration, continuous delivery, continuous deployment, github actions, ci, cd
```
The tag will be used to search and filter the content. People will generally search the content with space-separated words so I prefer to use either a single keyword or space-separated words.

banner: https://blog.jankaritech.com/src/assets/githubAction/images/githubaction_banner.png
---

> GitHub Actions is an automation and CI/CD service provided by GitHub. It allows us to automate our software development process or workflows, from building and testing code to deploying it to different environments. It's a really effective tool that can both increase software quality and save us time.

Reviewer 7: ```suggestion
GitHub Actions is an automation and CI/CD service provided by GitHub. It allows us to automate our software development process or workflows, from building and testing code to deploying it to different environments. It's a really an effective tool that can both increase software quality and save our time.
```

Reviewer 2: I think there are two ways how the last sentence can be expressed in English, either how Nabin first proposed it:

It's a really effective tool
--> meaning very effective

or 

It's really an effective tool
--> meaning it is true that this is an effective tool

> ## Some Features of GitHub Actions:

Reviewer 2: ```suggestion
## Some features of GitHub Actions:
```

Reviewer 2: Usually only the first letter of a sentence and names are capitalized in English. For titles, some people capitalize every word (apart from pronouns), but if you choose that style, apply it to the whole blog post.

> - It allows you to automate your Software Development Life Cycle (SDLC) processes or workflows.

Reviewer 6: Does GitHub Actions automate the whole SDLC??

Author: NO, the statement `It allows you to automate your Software Development Life Cycle (SDLC) processes or workflows` does not imply automating the entire SDLC process. 
Instead, it suggests that GitHub Actions enables automation for certain aspects or stages of the SDLC. The term `processes or workflows` indicates that specific tasks or workflows within the SDLC can be automated using GitHub Actions, but it does not necessary cover the complete SDLC.

I updated that point, making it clear and understandable so that others would not be confused. 
Thank you :hugs: for your review.

- It simplifies CI/CD integration, facilitating seamless DevOps practices.

## Components of GitHub Actions:
1. Workflows
2. Events
3. Jobs
4. Steps
5. Actions
6. Runners

![components](/src/assets/githubAction/images/components_of_github_action.png)

> ### 1. Workflows:

Reviewer 2: ```suggestion
### 1. Workflows
```
Sorry, I didn't notice this before. In English you usually use the double point if you have a list following (often not even formatted as a list) or if it is kind of like a subtitle, but not formatted as a title (like `Note:` on line 45)

Workflows are automated procedures that you add to your repository. They will run when triggered by an event in your repository, or they can be triggered manually, or at a defined schedule. They can be used to build, test, package, release, or deploy a project on GitHub.

> ### 2. Events:

Reviewer 2: ```suggestion
### 2. Events
```

> Events are specific activities like code pushes, pull requests creation, issue creation, issue comments, etc. that triggers a workflow.

Reviewer 7: ```suggestion
Events are specific activities like code push, pull requests creation, issue creation, comments on pull requests or issues, etc. that triggers a workflow.
```
if you mean like this

Reviewer 2: maybe put "pull request creation" also into singular (instead of pull request**s** creation) since the other items are also singular

### 6. Runner
A runner is a GitHub actions server where our workflows run. A runner listens for available jobs. Each runner runs multiple jobs at a time, and reports the progress, logs, and results back to GitHub. GitHub hosted runners are based on Ubuntu, Linux, Microsoft Windows, and Mac OS.

Note:
One event can trigger many workflows, a workflow can contain many jobs, and a job can contain many steps.

### Example:

![process_in_githubActions](/src/assets/githubAction/images/process.png)

## Getting started:

Before we dive into an example, make sure you have a repository where you want to set up your CI/CD workflow. If you don't have one, create a new repository for example named `action-hero`.

Step 1: Clone your repository into your system
```bash
git clone <your_github_repo_url>
```

Step 2: Go to path `action-hero` and make a folder `.github`
```bash
cd action-hero
mkdir .github
```

Step 4: Then, go to path `.github` and make a folder `workflows`
```bash
cd .github
mkdir workflows
```

Step 5: Inside the folder `workflows` create a `hello_world.yml` file
```bash
cd workflows
touch hello_world.yml
```

> Our file structure will look like this

Reviewer 7: ```suggestion
Our file structure should look like this
```

```

> 📦Action-hero

Reviewer 6: ```suggestion
📦action-hero
```

┗ 📂.github
┃ ┣ 📂workflows
┃ ┃ ┗ 📜hello_world.yml
```

Step 6: Add the following code snippet to your `hello_world.yml` file:

```
name: Hello world workflow

on:
  push:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    - name: Print a message
      run: echo "Hello, GitHub Actions!"
```

In this example:
- The workflow is triggered by a push event on the main branch.
- The job named "build" runs on the latest version of Ubuntu.
- The first step checks out the repository using the actions/checkout action.
- The second step simply prints a message using the echo command.

## GitHub Actions activity:

![github activity](/src/assets/githubAction/images/output.png)

Above image is an output from a GitHub Actions workflow run. The output shows that the workflow was successful and completed the `Hello world workflow` job, which consisted of two steps: `Checkout repository` and `Print a message`.

> The `Print a message` step executed the command echo "Hello, GitHub Actions!", resulting in the output `Hello, GitHub Actions!` being displayed in the workflow log.

Reviewer 3: ```suggestion
The `Print a message` step executed the command `echo "Hello, GitHub Actions!"`, resulting in the output `Hello, GitHub Actions!` being displayed in the workflow log.
```

## Conclusion:
In this blog, we've explored GitHub Actions, an automation and CI/CD service provided by GitHub. We've learned about its key components and how they work together to automate software development workflows.

GitHub Actions enables us to automate tasks like building, testing, and deploying code, leading to improved software quality and time savings.

With a practical example, we've seen how easy it is to set up a basic workflow.

> In summary, GitHub Actions is a powerful tool for speedy development processes, making it easier to deliver high-quality software efficiently.

Reviewer 7: ```suggestion
In summary, GitHub Actions is a powerful tool for speedy development processes, making it easier to deliver high-quality of software efficiently.
```

Reviewer 2: @SagarGi if "high-quality" is used as an attribute to specify the type/matter of software the preposition "of" is not needed. Alternatively you could say software of high-quality, but I think this expression is less common. Maybe ask @phil-davis or your English teacher :)

Reviewer: Correct. "high-quality software" is good.
"high-quality" is the adjective (describes an attribute of the software)
"software" is the noun (thing)

---
title: Github Actions - CI &CD
authorName: Nabin Ale
# authorAvatar : https://www.jankaritech.com/images/2023/06/06/nabin-800x800
authorAvatar : https://avatars.githubusercontent.com/u/61624650?s=400&u=8f992a9ca189af19c40cb070a5af1e02669d3884&v=4
authorLink: https://github.com/nabim777

> createdAt: Oct 30, 2023

Author: Okay @koebel I will change it at the time of merge. Will be it okay na?

tags: CI, CD, githubAction
banner: https://raw.githubusercontent.com/JankariTech/blog/master/src/imgs/fallback_banner.png
---
GitHub Actions is an automation and CI/CD service provided by GitHub. It allows us to automate our software development process or workflows, from building and testing code to deploying it to different environments. It's a really effective tool that can both increase software quality and save us time.

> ## Features in github actions:

Reviewer: Note: "GitHub" is the correct way to write the commercial name.
```suggestion
## Features in GitHub Actions:
```

> ## Features in github actions:

Author: > And by the way, you can also bring your blog post to English class and ask the English teacher for proof reading. She said to me that she is happy to help with such things :)

Okay @koebel

- automate SDLC process or workflow
- implemnts CI/CD DevOps

> 

Author: > are these all the features provided by GitHub Actions or just the ones you'll cover in the blog?
> 

I have just only write some features of GitHub actions only

> 

Reviewer 2: In that case it's better to write "some features of GitHub Action" or "Features of GitHub Action that will be covered in this blog" or something similar either in the heading of this section or as an introduction sentence.

Before we dive into an example, make sure we have a GitHub account and a repository where we want to set up your CI/CD workflow. If we don't have one, create a new repository.

## Components of github actions:
1. Workflows
2. Events
3. Jobs
4. Steps
5. Actions
6. Runners

> ![components](/src/assets/githubAction/components_of_github_action.png)

Author: I have created all the images I have used in this blog so no need for sources

Note:
One event can trigger many workflows, a workflow can contain many jobs, and a job can contain many steps

### Example:

![process_in_githubActions](/src/assets/githubAction/haha.png)

Before we dive into an example, make sure we have a repository where we want to set up your CI/CD workflow. If we don't have one, create a new repository.
Then create a folder `.github/workflows` and inside it make a file as `hello_world.yml`.

Add the following code snippet to your file `hello_world.yml`

```
name: Hello world workflow

on:
  push:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    - name: Print a message
      run: echo "Hello, GitHub Actions!"
```
In this example:
- The workflow is triggered by a push event on the main branch.
- The job named "build" runs on the latest version of Ubuntu.
- The first step checks out the repository using the actions/checkout action.

> - The second step simply prints a message using the echo command.

Author: ![process](https://github.com/JankariTech/blog/assets/61624650/82d8bbb7-9090-4a13-b8e5-e7915c417fc9)
Is it ok @koebel

> - The second step simply prints a message using the echo command.

Reviewer 2: looks very nice. just "issues" is still written with a small "i"

> - The second step simply prints a message using the echo command.

Author: Oh I forget that :sweat_smile:
![process](https://github.com/JankariTech/blog/assets/61624650/c01b6b54-f350-40d1-9667-93d90273f926)
Is that okay now @koebel ?

> - The second step simply prints a message using the echo command.

Reviewer 2: looks good
